[CostModel] Handle vector struct results and cost `llvm.sincos`

MacDue opened the pull request:
This patch updates the cost model to cost intrinsics that return multiple values (in structs) correctly. Previously, the cost model only thought intrinsics that return `VectorType` need scalarizing, which meant it cost intrinsics that return multiple vectors (that need scalarizing) way too cheap (giving it the cost of a single function call).

This patch also adds a custom cost for llvm.sincos when a vector function library is available, as certain VFs can be expanded (later in code gen) to a vector function, reducing the cost to a single call (+ the possible loads from the vector function returns values via output pointers).

MacDue commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -2005,10 +2049,13 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
-        ScalarizationCost += getScalarizationOverhead(
-            cast<VectorType>(RetTy),
-            /*Insert*/ true, /*Extract*/ false, CostKind);
+      if (!RetTy->isVoidTy()) {
+        for (Type *VectorTy : getContainedTypes(RetTy)) {
+          ScalarizationCost += getScalarizationOverhead(
+              cast<VectorType>(VectorTy),
+              /*Insert*/ true, /*Extract*/ false, CostKind);
+        }
+      }
```
@arsenm this is what causes the changes in the AMDGPU `frexp.ll` costs. Previously, this would be skipped as `RetVF` was 1 for struct-of-vector returns. Now it costs the inserts for both returned vectors.

MacDue commented on llvm/test/Transforms/Scalarizer/sincos.ll:
```diff
@@ -1,17 +0,0 @@
-; CHECK-NEXT:    [[EL:%.*]] = extractvalue { <4 x float>, <4 x float> } [[R]], 0
-; CHECK-NEXT:    ret <4 x float> [[EL]]
-;
-  %r = call { <4 x float>, <4 x float> } @llvm.sincos.v4f32(<4 x float>  %Val)
-  %el = extractvalue { <4 x float>, <4 x float> } %r, 0
```
Note: I replaced this with an equivalent `llvm.vector.deinterleave2` to test (which matches the intent of this test -- checking non-trivially-scalarizable struct-ret intrinsics are not scalarized).

SamTebbs33 reviewed (approved):
This all looks sensible to me.

huntergr-arm commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -285,6 +286,64 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
+  /// the call return value. If std::nullopt is returned, the no vector library
+  /// call is available, so the intrinsic should be assigned the default cost
+  /// (e.g. scalarization).
+  std::optional<InstructionCost> getMultipleResultIntrinsicVectorLibCallCost(
```
I think returning a cost of 'Invalid' should suffice, rather than adding another layer?

MacDue replied:
Prefer `std::optional` as it does not overload the meaning of 'Invalid' for "no cost decided". You could accidentally return the 'Invalid' cost as the final cost, whereas an `std::optional` forces you to check it.

huntergr-arm commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -285,6 +286,64 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
+          VectorType::get(IntegerType::getInt1Ty(Ctx), VF), {}, CostKind, 0,
+          nullptr, {});
+
+    // Lowering to a library call (with output pointers) may require us to emit
```
Always adding on the cost of a load seems pessimistic?

MacDue commented on llvm/test/Transforms/LoopVectorize/AArch64/llvm.sincos.ll:
```diff
@@ -0,0 +1,170 @@
+; CHECK-COST: Cost of Invalid for VF vscale x 2: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
+; CHECK-COST: Cost of Invalid for VF vscale x 4: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
+
+; CHECK-COST-ARMPL-LABEL: sincos_f32
+; CHECK-COST-ARMPL: LV: Found an estimated cost of 10 for VF 1 For instruction:   %call = tail call { float, float } @llvm.sincos.f32(float %in_val)
+; CHECK-COST-ARMPL: Cost of 26 for VF 2: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
+; CHECK-COST-ARMPL: Cost of 12 for VF 4: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
+; CHECK-COST-ARMPL: Cost of Invalid for VF vscale x 1: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
+; CHECK-COST-ARMPL: Cost of Invalid for VF vscale x 2: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
+; CHECK-COST-ARMPL: Cost of 13 for VF vscale x 4: WIDEN-INTRINSIC ir<%call> = call llvm.sincos(ir<%in_val>)
```
Done :+1: I did need to slightly tweak the cost model pass to forward the library info.

david-arm commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -285,6 +286,64 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
+
+    // Find associated libcall.
+    const char *LCName = getTLI()->getLibcallName(LC);
+    if (!LC || !LCName)
```
Hmm, I'm not sure I understand why you check for `!LC` here because LC=0 seems to be a valid enum. See llvm/include/llvm/IR/RuntimeLibcalls.h:

```
enum Libcall {
#define HANDLE_LIBCALL(code, name) code,
#include "llvm/IR/RuntimeLibcalls.def"
#undef HANDLE_LIBCALL
};
```

where the first entry in `llvm/include/llvm/IR/RuntimeLibcalls.def` is:

```
// Integer
HANDLE_LIBCALL(SHL_I16, "__ashlhi3")
```

MacDue replied:
Good spot! Removed the erroneous !LC check :+1:

david-arm commented on llvm/test/Analysis/CostModel/AArch64/sincos.ll:
Should we also add tests for the masked variants too?

david-arm commented on llvm/test/Analysis/CostModel/AArch64/sincos.ll:
```diff
@@ -0,0 +1,67 @@
+  %f128 = call { fp128, fp128 } @llvm.sincos.f128(fp128 %x_f128)
+
+  %v8f16 = call { <8 x half>, <8 x half> } @llvm.sincos.v8f16(<8 x half> %x_v8xf16)
+  %v4f32 = call { <4 x float>, <4 x float> } @llvm.sincos.v4f32(<4 x float> %x_v4xf32)
```
It would be good to have at least one test for wider vectors that would need splitting too, i.e.

```
%v8f32 = call { <8 x float>, <8 x float> } @llvm.sincos.v8f32(<8 x float> %x_v8xf32)
```

and similar with something like `<vscale x 8 x float>`

david-arm reviewed (commented):
Given you've got explicit cost model tests in llvm/test/Analysis/CostModel/AArch64/sincos.ll for the BasicTTIImpl changes I wonder if it's worth pulling the vectoriser changes out into a follow-on PR? That way you at least get to keep most of the changes upstream if the vectoriser bits get reverted for some reason.

MacDue committed:
[CostModel] Handle vector struct results and cost `llvm.sincos`

This patch updates the cost model to cost intrinsics that return
multiple values (in structs) correctly. Previously, the cost model only
thought intrinsics that return `VectorType` need scalarizing, which
meant it cost intrinsics that return multiple vectors (that need
scalarizing) way too cheap (giving it the cost of a single function
call).

This patch also adds a custom cost for llvm.sincos when a vector
function library is available, as certain VFs can be expanded
(later in code gen) to a vector function, reducing the cost to a
single call (+ the possible loads from the vector function returns
values via output pointers).

MacDue commented:
> I wonder if it's worth pulling the vectoriser changes out into a follow-on PR? That way you at least get to keep most of the changes upstream if the vectoriser bits get reverted for some reason.

Sure, I've moved the LV changes to https://github.com/llvm/llvm-project/pull/128035/commits/21b534d9cd8b0a5b7c99272b64e59a9f3fa9d5e6.

MacDue commented on llvm/test/Analysis/CostModel/AArch64/sincos.ll:
```diff
@@ -0,0 +1,67 @@
+; CHECK-VECLIB:  Cost Model: Found an estimated cost of 13 for instruction: %nxv2f64 = call { <vscale x 2 x double>, <vscale x 2 x double> } @llvm.sincos.nxv2f64(<vscale x 2 x double> %x_nxv2xf64)
+; CHECK-VECLIB:  Cost Model: Invalid cost for instruction: %nxv1f128 = call { <vscale x 1 x fp128>, <vscale x 1 x fp128> } @llvm.sincos.nxv1f128(<vscale x 1 x fp128> %x_nxv1xf128)
+
+  half %x_f16,
```
Ah yeah, I forgot about `poison` (I didn't use `undef` like the other tests as that'd upset the new bot).

MacDue committed:
Update tests

arsenm commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -2005,10 +2074,13 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
+        for (Type *VectorTy : getContainedTypes(RetTy)) {
+          ScalarizationCost += getScalarizationOverhead(
+              cast<VectorType>(VectorTy),
+              /*Insert*/ true, /*Extract*/ false, CostKind);
```
```suggestion
              /*Insert=*/ true, /*Extract=*/ false, CostKind);
```

arsenm commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -2689,27 +2761,32 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
+        ScalarizationCost = 0;
+        for (Type *RetVTy : RetVTys) {
+          ScalarizationCost += getScalarizationOverhead(
+              cast<VectorType>(RetVTy), /*Insert*/ true,
+              /*Extract*/ false, CostKind);
```
```suggestion
              cast<VectorType>(RetVTy), /*Insert=*/ true,
              /*Extract=*/ false, CostKind);
```

arsenm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -69,9 +69,9 @@ bool HardwareLoopInfo::canAnalyze(LoopInfo &LI) {
 IntrinsicCostAttributes::IntrinsicCostAttributes(
     Intrinsic::ID Id, const CallBase &CI, InstructionCost ScalarizationCost,
-    bool TypeBasedOnly)
+    bool TypeBasedOnly, TargetLibraryInfo const *LibInfo)
```
```suggestion
    bool TypeBasedOnly, const TargetLibraryInfo *LibInfo)
```

arsenm commented on llvm/lib/Analysis/CostModel.cpp:
```diff
@@ -42,25 +44,31 @@ static cl::opt<bool> TypeBasedIntrinsicCost("type-based-intrinsic-cost",
     cl::desc("Calculate intrinsics cost based only on argument types"),
     cl::init(false));
 
+static cl::opt<bool> LibCallBasedIntrinsicCost(
+    "libcall-based-intrinsic-cost",
+    cl::desc("Calculate intrinsics cost using target library info"),
+    cl::init(false));
```
Don't understand why this would be an option Remove this?

MacDue replied:
Explained below https://github.com/llvm/llvm-project/pull/123210#discussion_r1965022581

arsenm commented on llvm/lib/Analysis/CostModel.cpp:
```diff
@@ -42,25 +44,31 @@ static cl::opt<bool> TypeBasedIntrinsicCost("type-based-intrinsic-cost",
-        IntrinsicCostAttributes ICA(II->getIntrinsicID(), *II,
-                                    InstructionCost::getInvalid(), true);
+      if (II && (LibCallBasedIntrinsicCost || TypeBasedIntrinsicCost)) {
+        IntrinsicCostAttributes ICA(
+            II->getIntrinsicID(), *II, InstructionCost::getInvalid(),
+            /*TypeBasedOnly=*/TypeBasedIntrinsicCost, &TLI);
```
You don't know if this is a libcall or not here, I also don't understand why this is guarded by TypeBasedIntrinsicCost. Can we remove that one too?

MacDue replied:
I only use the `getIntrinsicInstrCost` code path if one of the flags (`type-based-intrinsic-cost` or `libcall-based-intrinsic-cost`) is set as I discovered some targets return a different cost depending on if you call `getIntrinsicInstrCost` directly or via `getInstructionCost`. This seemed to be due to them modifying the cost returned from `getIntrinsicInstrCost` within their implementation of `getInstructionCost`. I didn't want to change tests that depend on this (and it's not relevant for AArch64), so I added a flag.

david-arm replied:
I must admit though it does seem a little odd now because `-type-based-intrinsic-cost` also implies libcall-based-intrinsic-cost since it causes `TLI` to be passed in. How about renaming `libcall-based-intrinsic-cost` to be `prefer-intrinsic-cost` or `use-intrinsic-cost`, since it's not really tied to the library calls?

This is just a suggestion, but if we had a `use-intrinsic-cost` option we could actually make it an enum along the lines of:

```
enum IntrinsicCostType {
  None,
  NonTypeBased,
  TypeBased,
};
```

that way you can collapse the two options into one and both `TypeBased` and `NonTypeBase` variants will use the TLI. However, I appreciate that would significantly increase the number of test files changed so it's possibly something for a follow-on?

MacDue replied:
I've renamed the flag `prefer-intrinsic-cost` for this PR. The enum works nicely :+1:, but since it does increase the churn a bit I agree it's best to post that as a small follow-up PR.

MacDue replied:
Just to be clear this flag only affects the test cost-model print pass, not LLVM more generally. The loop vectorizer in https://github.com/llvm/llvm-project/pull/128035 always passes the TLI.

MacDue replied:
Changing the default would not be much trouble though -- it just requires adding a flag to a few tests.

david-arm replied:
The problem I see with changing the default `prefer-intrinsic-cost` to true is that some tests may genuinely want to exercise the `getInstructionCost` path, in which case they have to live without using the TLI. I think it makes sense to follow this up in a separate PR, along with potentially changing it to use an enum instead of a boolean so that we can remove the `type-based-intrinsic-cost` flag as well?

MacDue replied:
Yep, I'm happy to address this in the follow-up PR (since there will be a little test churn in that patch anyway).

fhahn replied:
> Just to be clear this flag only affects the test cost-model print pass (so users = LLVM developers), not LLVM more generally. The loop vectorizer in https://github.com/llvm/llvm-project/pull/128035 always passes the TLI.

Ok that's good, thanks. I think ideally we would get rid of the option as follow-up, as it seems confusing to have the cost model printer show something that's different to the cost that is actually used in the passes making the queries.

MacDue committed:
Fixups

MacDue commented:
Would be good to get this in soon, as I have some downstream changes depending on it :)

david-arm commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -285,6 +286,64 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
+  /// the cost, which includes the cost of the mask (if required), and the loads
+  /// for values returned via output pointers. \p LC is the scalar libcall and
+  /// \p CallRetElementIndex (optional) is the struct element which is mapped to
+  /// the call return value. If std::nullopt is returned, the no vector library
```
nit: I think this should be `then no vector ...`

MacDue committed:
Fixups

david-arm reviewed (approved):
LGTM, but please wait a day in case others have any objections!

MacDue commented:
Thanks for the reviews! I've got the final patch for `llvm.sincos` here https://github.com/llvm/llvm-project/pull/128035, which is a trivial patch which just enables it's vectorization.

MacDue commented:
Created a follow-up for the flags here: https://github.com/llvm/llvm-project/pull/128885. I've not removed them or changed the default yet, but feel free to discuss that on the new patch.
